web/src/Web.App/Services/FinanceService.cs

```diff
@@ -9,7 +12,34 @@ public interface IFinanceService
     Task<FinanceYears> GetYears();
 }
 
-public class FinanceService(IInsightApi insightApi) : IFinanceService
+public class FinanceService(
+    IInsightApi insightApi,
+    IMemoryCache memoryCache,
+    IOptions<CacheOptions> cacheOptions) : IFinanceService
 {
-    public async Task<FinanceYears> GetYears() => await insightApi.GetCurrentReturnYears().GetResultOrThrow<FinanceYears>();
-}
\ No newline at end of file
+    private readonly CacheOptions _cacheOptions = cacheOptions.Value;
+    private readonly string _cacheKey = cacheOptions.Value.CacheKey ?? throw new NullReferenceException(nameof(CacheOptions));
+
+    public async Task<FinanceYears> GetYears()
+    {
+        FinanceYears data;
+        if (memoryCache.TryGetValue(_cacheKey, out var cachedData))
+        {
+            data = cachedData as FinanceYears ?? throw new InvalidOperationException(nameof(FinanceYears));
```

Comment:
If the cached data is not of the expected format it would be better to just get and cache a new value rather than throw an exception. A warning log message may be helpful though.

Reply:
thanks, updated to do this in latest commits